### ISSUE TITLE
Flow: otelcol.processor.attributes component

### ISSUE THREAD

**Issue by rfratto** [grafana/agent#2294]
Expose the OpenTelemtry Collector `attributes` processor as a Flow component called `otelcol.processor.attributes`. This component is from the `otelcol-contrib` distribution.

**Comment by ptodev** [comment]
I got stuck on this project because I need to create instances of "filterconfig.MatchConfig" and "attraction.Settings" but they are internal:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/attributesprocessor/config.go#L37

One way to do this is to fork the contrib repo and export all the internals, as we currently do with the non-contrib one. @gouthamve suggested to try exporting those and checking if the changes get accepted in the official contrib repo before we fork. 

I forked the contrib repo in my private GitHub area and exported those in the pkg folder. This is on tag [v0.0.900](https://github.com/ptodev/opentelemetry-collector-contrib/tree/v0.0.900). However, then the Agent didn't compile because it was using an old version of OTEL which has a function FromRaw which does not return an error. FromRaw was changed to return an error [here](https://github.com/open-telemetry/opentelemetry-collector/pull/6579) at version [0.65.0](https://github.com/open-telemetry/opentelemetry-collector/releases/tag/v0.65.0) of OTEL. I tried updating the version of OTEL which the Agent uses, but then there were other compilation problems.

Finally, I tried rolling back both my OTEL and contrib changes, and using version 0.61.0 of otel-contrib. But it turns out that this version has the filter library in an old, awkward location which would look strange if exported.

It seems to me that the only realistic way to move forward is to fork the contrib repo and export all internals. @tpaschalis mentioned that the exporting script we use for OTEL is [this one](https://github.com/grafana/opentelemetry-collector/blob/0.67-grafana/export-internal.sh), so maybe we can have some variation of this for contrib.

The main thing that bothers me about the forks is that OTEL could one day remove config attributes, or even change the meaning of some attributes. I guess in those cases clients would not be able to upgrade to new versions of the Agent until they change their OTEL config, and there would be only one version of OTEL that a given version of Agent supports.

By the way, while speaking with @kovrus he mentioned that a lot of OTEL components (such as the attributes processor) are implemented using the [OTTL language](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl). We could theoretically also use this language, although at the moment I guess we are just writing adaptors to already existing OTEL components. But what worries me about OTTL is that people might have a complicated OTTL config which they try to use within the Agent - then there would be a mixture of OTTL and River and that may look complicated.

**Comment by ptodev** [comment]
I managed to build a fork of otel-contrib based on the 0.63.0 branch where the "internal" folder is renamed to "external":
https://github.com/ptodev/opentelemetry-collector-contrib/tree/external-v.0.63.0-attempt3

I also made a lot of progress with the attribute processor code - it's on the branch indicated on this issue. It's still not finished though... I will try to finish it when I get back in January.

It would be good of we can decide how we want to import otel-contrib. We could use fork where we make the internals external. But a cleaner approach could be to translate River to Yaml, and then use factories in the OTEL code which convert Yaml to the OTEL data structure. Then we can just use the data structure from there. If there are no such factories already, I think we could add them - such a change would be likely to be accepted to the OTEL repo? Another upside to this approach is that it would allow us to test the translation from River to Otel more easily. We could use configs like [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/attributesprocessor/testdata/config.yaml) in our unit tests to see if our River inputs match them.

I think we were also discussing that a translation between River and Yaml might be a requirement for the Agent to become a distribution, so it may be something we want to do anyway.

That said, a fork might be the only realistic short term solution.

**Comment by erikbaranowski** [comment]
@rfratto suggested a workaround that did in fact work for this issue (thank you!). We used the otel mapstructure to decode our river types into otel types. There is some example code here showing the implementation.

[tail_sampling.go](https://github.com/grafana/agent/blob/main/component/otelcol/processor/tail_sampling/tail_sampling.go)
[types.go](https://github.com/grafana/agent/blob/main/component/otelcol/processor/tail_sampling/types.go)

As a second optional consideration, I implemented the squashed otel type sharedPolicyCfg by skipping it in the river types and using good old copy and paste in the 3 river types that use it. You have this option if you don't want to have to touch user docs after river implements squashing. In short, the downside is the repeated code in types.go and the upside is not having to change the river config/docs later when river implements squashing.

**Comment by rfratto** [comment]
The overall [Config type](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/92ad54f707a967386b764b6fdaea5a2ba4377319/processor/attributesprocessor/config.go#L31) from the processor is exported, so we can do mapstructure.Decode into it. 

> I doubt it would be able to do this when the Otel types are internal.

It can. mapstructure.Decode is how OpenTelemetry Collector decodes YAML configs, which is why OpenTelemetry Collector can create a config for that copmonent even though it has types from internal packages.

**Comment by ptodev** [comment]
Thanks @rfratto. I suspected that this works for Otel because these internal types are available to them, and they call `mapstructure` from within Otel. But the types are not available to us, hence why I thought it wouldn't work. 

Just now I tried this in a unit test and it built ok:
```
func Test(t *testing.T) {
	input := map[string]interface{}{
		"actions": []interface{}{
			map[string]interface{}{
				"key":    "attribute1",
				"value":  123,
				"action": "insert",
			},
		},
	}

	var result attributesprocessor.Config
	err := mapstructure.Decode(input, &result)
	require.NoError(t, err)
}
```

In the above unit test I tried adding this line:
```
require.Equal(t, "insert", result.Actions[0].Action)
```

The test failed with this error:
```
Error:      	Not equal: 
            	expected: string("insert")
           	actual  : attraction.Action("insert")
```

When tried this instead, it wouldn't compile because it can't find the attraction package (I suppose because it's internal to Otel):
```
require.Equal(t, attraction.Action("insert"), result.Actions[0].Action)
```

So I basically just used a `mapstructure.Decode` call to get an object from within the Otel, which is internal. I can't even initialise it on my own if I wanted. Not sure how it's possible... I can't wrap my mind around this 🤯  Anyway, I'll try to write something up to create an attributesprocessor.Config in a similar way. Thanks for the help!

**Comment by rfratto** [comment]
I took a quick look at it, and yes, that approach is the idea for configuring OpenTelemetry components that are using internal packages.

**Comment by zackman0010** [comment]
@rfratto Can this ticket be reopened for tracking due to the code being reverted with #3589?

**Comment by ptodev** [comment]
This has been merged to `main` via #3822
